core/blockchain/block_tree.hpp

```diff
@@ -0,0 +1,118 @@
+   *      forks, pruning the blocks, resolving child-parent relations, etc)
+   */
+  struct BlockTree {
+    using BlockHashVecRes = outcome::result<std::vector<primitives::BlockHash>>;
+
+    virtual ~BlockTree() = default;
+
+    /**
+     * Get a body (extrinsics) of the block
+     * @param block - id of the block to get body for
+     * @return body, if the block exists in our storage, error otherwise
+     */
+    virtual outcome::result<primitives::BlockBody> getBlockBody(
+        const primitives::BlockId &block) const = 0;
+
+    /**
+     * Add a new block to the tree
+     * @param block to be added
+     * @return nothing or error; if error happens, no changes in the tree are
+     * made
+     *
+     * @note if block, which is specified in PARENT_HASH field of (\param block)
+     * is not in our local storage, corresponding error is returned. It is
+     * suggested that after getting that error, the caller would ask another
+     * peer for the parent block and try to insert it; this operation is to be
+     * repeated until a successful insertion happens
+     */
+    virtual outcome::result<void> addBlock(primitives::Block block) = 0;
+
+    /**
+     * Mark the block as finalized and store a finalization justification
+     * @param block to be finalized
+     * @param justification of the finalization
+     * @return nothing or error
+     */
+    virtual outcome::result<void> finalize(
+        const primitives::BlockHash &block,
+        const primitives::Justification &justification) = 0;
+
+    /**
+     * Get a chain of blocks from the specified as a (\param block) up to the
+     * closest finalized one
+     * @param block to get a chain from
+     * @return chain of blocks or error
+     */
+    virtual BlockHashVecRes getChainByBlock(
+        const primitives::BlockHash &block) = 0;
+
+    /**
+     * Get a longest path (chain of blocks) from the last finalized block down
+     * to the deepest leaf
+     * @return chain of blocks or error
+     *
+     * @note this function is equivalent to "getChainByBlock(deepestLeaf())"
+     */
+    virtual BlockHashVecRes longestPath() = 0;
+
+    /**
+     * Get a deepest leaf of the tree
+     * @return deepest leaf
+     *
+     * @note deepest leaf is also a result of "SelectBestChain": if we are the
+     * leader, we connect a block, which we constructed, to that deepest leaf
+     */
+    virtual const primitives::BlockHash &deepestLeaf() const = 0;
+
+    /**
+     * Get all leaves of our tree
+     * @return collection of the leaves
+     */
+    virtual std::vector<primitives::BlockHash> getLeaves() const = 0;
+
+    /**
+     * Get children of the block with specified hash
+     * @param block to get children of
+     * @return collection of children hashes or error
+     */
+    virtual BlockHashVecRes getChildren(const primitives::BlockHash &block) = 0;
+
+    /**
+     * Get a last finalized block
```

Comment:
If there is only one last finalized block, which may be important,
```suggestion
     * Get the last finalized block
```